### ISSUE TITLE
AspNetCore authorization scopes policies helper

### ISSUE THREAD

**Issue by baywet** [AzureAD/microsoft-authentication-library-for-dotnet#1158]
**Is your feature request related to a problem? Please describe.**
If you're building an aspnetcore mvc application [Microsoft.AspNetCore.Authentication.AzureAD.UI](https://github.com/aspnet/AspNetCore/tree/master/src/Azure/AzureAD) does a fantastic job at seemlessly integrating your application with AAD from an authN standpoint.

Now let's assume you also want to build an API, and that API will have different surfaces that are secured using scopes. (ie: I don't want to grant access to all applications to the complete API endpoints exposed by my API).

Because AAD appends all the scopes as "scope1 scope2 scope3" in the scp claim instead of using a multi-valued claim ( scopes: ["scope1", "scope2", "scope3"]) like other authentication providers do, it's not possible to leverage existing scope policies libraries out there. (like identityserver4.accesstokenvalidation which provides ScopePolicy.Create).
It is also not possible to leverage OOTB aspnetcore APIs like a policy builder with a claim requirement.

**Describe the solution you'd like**
It'd be nice to have new package Microsoft.AspNetCore.Authorization.AzureAD, this package would basically contain two static mehods:
```CSharp
namespace Microsoft.AspNetCore.Authentication {
public static class AzureADAuthorizationBuilderExtensions {
public static AuthorizationBuilder AddAzureAD(this AuthorizationBuilder, string scheme...); // this method would add the AuthorizationHandler to the services
}
}
```
```CSharp
namespace Microsoft.AspNetCore.Authorization
{
    //
    // Summary:
    //     Helper for creating scope-related policies
    public static class ScopePolicy
    {
        //
        // Summary:
        //     Creates a policy to check for required scopes.
        //
        // Parameters:
        //   scopes:
        //     List of any required scopes. The token must contain at least one of the listed
        //     scopes.
        public static AuthorizationPolicy Create(params string[] scopes);
    }
}
```
Here is some example handler
```CSharp
public class OneOfTheScopesHandler : AuthorizationHandler<OneOfTheScopesRequirement>
    {
        private const string scopeName = "http://schemas.microsoft.com/identity/claims/scope";
        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, OneOfTheScopesRequirement requirement)
        {
            if (!context.User.HasClaim(x => x.Type.Equals(scopeName)))
            {
                context.Fail();
            }
            else
            {
                var currentScopes = context.User.Claims.Where(x => x.Type.Equals(scopeName))
                    .Select(x => x.Value)
                    .SelectMany(x => x.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries));
                if (currentScopes.Any(x => requirement.Scopes.Any(y => y.Equals(x, StringComparison.InvariantCultureIgnoreCase))))
                    context.Succeed(requirement);
                else context.Fail();
            }
            return Task.CompletedTask;
        }
    }
public class OneOfTheScopesRequirement: IAuthorizationRequirement
    {
        public OneOfTheScopesRequirement(params string[] scopes)
        {
            Scopes = scopes;
        }
        public string[] Scopes { get; set; }
    }
```

**Additional context**
I think all of that should seat under the aspnetcore repo I pointed to earlier but should be owned by the same team owning MSAL. So before starting any work I wanted to get the conversation going.

Having @Tratcher (https://github.com/Tratcher) part of the conversation could help as well.

**Comment by Tratcher** [comment]
@haok for Authorization

**Comment by jmprieur** [comment]
@Tratcher : do we want to move this issue to the ASP.NET Core repo? It's not related to MSAL.NET ?

**Comment by jmprieur** [comment]
cc: @MarkZuber FYI

**Comment by baywet** [comment]
@jmprieur the reason I started the conversation here is because the way of AAD of representing the scopes in a token is not "standard" (although the way to represent scopes doesn't seem to be part of the original JWT spec). IMHO the end library would belong to the aspnet core repo but I thought people here might be interested in having a conversation before starting any work.

**Comment by jmprieur** [comment]
Closing as a dupe of aspnet/AspNetCore#10447
